Simplify promise and completion handler management in FullscreenManager

achristensen07 opened the pull request:
#### 51b275a9492afcba8ee8f146ca80bbbd9ecaac77
<pre>
Simplify promise and completion handler management in FullscreenManager
<a href="https://bugs.webkit.org/show_bug.cgi?id=287531">https://bugs.webkit.org/show_bug.cgi?id=287531</a>
<a href="https://rdar.apple.com/144654454">rdar://144654454</a>

Reviewed by Jer Noble.

This simplifies logic flow in several tangled pieces of code.

The first simplification is in FullscreenManager::requestFullscreenForElement which used to
take an optional DeferredPromise and an optional CompletionHandler&lt;void(bool)&gt;.
To fix this, I moved the DeferredPromise handling to the one call site that
provided it, Element::requestFullscreen. I changed the CompletionHandler&lt;void(bool)&gt;
to take an ExceptionOr&lt;void&gt; instead of a bool to have enough information to reject
the promise with the right exception when that happens.  To retain the current
behavior of immediately rejecting the promise when another request or exit
fullscreen call happens, I continue to store a promise wrapper on the FullscreenManager,
but I store it as a CompletionHandler instead of a DeferredPromise.  Because of this,
I needed to do a similar transformation to DocumentFullscreen::exitFullscreen.
Now, they both store the CompletionHandler in m_pendingPromise and keep a WeakPtr
that will resolve the promise if it hasn&apos;t been resolved by the time the async
task is complete.  I introduced FullscreenPromise for this purpose.

In order to do this first simplification, I needed to change the time at which
the completion handler is called in the success case.  It used to be called immediately
after the call to page-&gt;chrome().client().enterFullScreenForElement(element, mode)
and now it is called after the reply to Messages::WebFullScreenManagerProxy::EnterFullScreen
is received.  This is not a change in behavior because the only callers of
FullscreenManager::requestFullscreenForElement which passed in a CompletionHandler were
HTMLMediaElement::enterFullscreen which only did anything in the failure case and
WebFullScreenManager::requestRestoreFullScreen which was only called from IPC from
WKFullScreenWindowController.requestRestoreFullScreen: which also only did anything
in the failure case, so changing the time at which the success case called its completion
handler doesn&apos;t change any behavior.

The second simplification is that FullscreenManager::willEnterFullscreen used to
return a bool and resolve or reject the stored promise.  To make this responsibility
more clear and avoid mistakes of forgetting one or the other, I pass in the completion
handler and that will take care of both when called.

* Source/WebCore/dom/DocumentFullscreen.cpp:
(WebCore::DocumentFullscreen::exitFullscreen):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::requestFullscreen):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::~FullscreenManager):
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::resolvePendingPromise):
(WebCore::FullscreenManager::rejectPendingPromise):
(WebCore::FullscreenManager::clear):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::enterFullscreen):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::willEnterFullScreen):
(WebKit::WebFullScreenManager::requestRestoreFullScreen):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::enterFullScreenForElement):

Canonical link: <a href="https://commits.webkit.org/290289@main">https://commits.webkit.org/290289@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e16dcc87d636178142430647344b905259e51d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89568 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44427 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94560 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40335 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91620 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9484 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17374 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26641 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92569 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49355 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39441 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/36681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96388 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16750 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77862 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17005 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/77095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/77178 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21609 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9917 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14049 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16763 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/16504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/18286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -248,40 +216,33 @@ void FullscreenManager::requestFullscreenForElement(Ref<Element>&& element, RefP
+            page->chrome().client().enterFullScreenForElement(element, mode, [weakPromise = WeakPtr { m_pendingPromise }] (ExceptionOr<void> result) {
+                if (weakPromise)
+                    weakPromise->rejectOrResolve(WTFMove(result));
+            });
```
I don't really understand the benefit of making `FullscreenPromise` a `CanMakeWeakPtr` when the manager itself is already a `CanMakeWeakPtr` and is passed into this lambda as a `WeakPtr` itself. Especially since the lifetime of a `FullscreenPromise` is tied to the lifetime of the `FullscreenManager` itself.

achristensen07 replied:
Before this change, there was a possibility that enterFullScreen would start, then exitFullScreen would start and reject the promise, then enterFullScreen would complete and reject the wrong promise.  Keeping a WeakPtr to the promise itself prevents this possible confusion.

achristensen07 replied:
I didn't actually fix this problem, so I'm going with your suggestion.  It is better.

jernoble commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -435,10 +384,13 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
+                page->chrome().client().enterFullScreenForElement(*m_pendingFullscreenElement, HTMLMediaElementEnums::VideoFullscreenModeStandard, [weakPromise = WeakPtr { m_pendingPromise }] (auto result) {
+                    if (weakPromise)
+                        weakPromise->rejectOrResolve(WTFMove(result));
+                });
```
Ditto here.

jernoble commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -819,6 +750,46 @@ WTFLogChannel& FullscreenManager::logChannel() const
+    clear();
+}
+
+auto FullscreenManager::FullscreenPromise::operator=(CompletionHandler<void(ExceptionOr<void>)>&& promise) -> FullscreenPromise&
```
The name of this class is weird, as is the name of the parameter, and the ivar. None of them is a promise; the parameter and the ivar are both `CompletionHandler`s, not Promises.

achristensen07 replied:
I agree it's a strange name.  It rejects and resolves like a promise, and it often wraps a promise.  It'll hopefully be soon removed.

jernoble reviewed (approved):
r=me, with nits.

achristensen07 committed:
Simplify promise and completion handler management in FullscreenManager
https://bugs.webkit.org/show_bug.cgi?id=287531
rdar://144654454

Reviewed by Jer Noble.

This simplifies logic flow in several tangled pieces of code.

The first simplification is in FullscreenManager::requestFullscreenForElement which used to
take an optional DeferredPromise and an optional CompletionHandler<void(bool)>.
To fix this, I moved the DeferredPromise handling to the one call site that
provided it, Element::requestFullscreen. I changed the CompletionHandler<void(bool)>
to take an ExceptionOr<void> instead of a bool to have enough information to reject
the promise with the right exception when that happens.  To retain the current
behavior of immediately rejecting the promise when another request or exit
fullscreen call happens, I continue to store a promise wrapper on the FullscreenManager,
but I store it as a CompletionHandler instead of a DeferredPromise.  Because of this,
I needed to do a similar transformation to DocumentFullscreen::exitFullscreen.
Now, they both store the CompletionHandler in m_pendingPromise and keep a WeakPtr
that will resolve the promise if it hasn't been resolved by the time the async
task is complete.  I introduced FullscreenPromise for this purpose.

In order to do this first simplification, I needed to change the time at which
the completion handler is called in the success case.  It used to be called immediately
after the call to page->chrome().client().enterFullScreenForElement(element, mode)
and now it is called after the reply to Messages::WebFullScreenManagerProxy::EnterFullScreen
is received.  This is not a change in behavior because the only callers of
FullscreenManager::requestFullscreenForElement which passed in a CompletionHandler were
HTMLMediaElement::enterFullscreen which only did anything in the failure case and
WebFullScreenManager::requestRestoreFullScreen which was only called from IPC from
WKFullScreenWindowController.requestRestoreFullScreen: which also only did anything
in the failure case, so changing the time at which the success case called its completion
handler doesn't change any behavior.

The second simplification is that FullscreenManager::willEnterFullscreen used to
return a bool and resolve or reject the stored promise.  To make this responsibility
more clear and avoid mistakes of forgetting one or the other, I pass in the completion
handler and that will take care of both when called.

* Source/WebCore/dom/DocumentFullscreen.cpp:
(WebCore::DocumentFullscreen::exitFullscreen):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::requestFullscreen):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::~FullscreenManager):
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::resolvePendingPromise):
(WebCore::FullscreenManager::rejectPendingPromise):
(WebCore::FullscreenManager::clear):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::enterFullscreen):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::willEnterFullScreen):
(WebKit::WebFullScreenManager::requestRestoreFullScreen):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::enterFullScreenForElement):

Canonical link: https://commits.webkit.org/290289@main
